src/ports/postgres/modules/graph/sssp.py_in

```diff
@@ -198,110 +378,228 @@ def graph_sssp(schema_madlib, vertex_table, vertex_id, edge_table,
-	ret = [dest_vertex]
-	plan_name = unique_string(desp='plan')
-
-	# Follow the 'parent' chain until you reach the source.
-	# We don't need to know what the source is since it is the only vertex with
-	# itself as its parent
-	plpy.execute(""" PREPARE {plan_name} (int) AS
-		SELECT parent FROM {sssp_table} WHERE {id} = $1 LIMIT 1
-		""".format(**locals()))
 	sql = "EXECUTE {plan_name} ({cur})"
 	parent = plpy.execute(sql.format(**locals()))
+	if parent.nrows() > 0:
+		while 1:
+			parent = parent[0]['parent']
+			if parent == cur:
+				ret.reverse()
+				ret = ",".join(ret)
+				plpy.execute(
+					""" INSERT INTO {path_table} VALUES ({gset} '{ret}')
+					""".format(**locals()))
+				return True
+			else:
+				ret.append(str(parent))
+				cur = parent
+
+			parent = plpy.execute(sql.format(**locals()))
+
+	return False
+
+def graph_sssp_get_path(schema_madlib, sssp_table, dest_vertex, path_table,
+	**kwargs):
+	"""
+    Helper function that can be used to get the shortest path for a vertex
+    Args:
+        @param source_table Name of the table that contains the SSSP output.
+        @param out_table    The vertex that will be the destination of the
+                            desired path.
+        @param path_table   Name of the output table that contains the path.
+	"""
+	with MinWarning("warning"):
+
+		_validate_get_path(sssp_table, dest_vertex)
+		plan_name = unique_string(desp='plan')
+
+		select_grps = ""
+		gsql = ""
+
+		path_flag = False
+
+		summary = plpy.execute("SELECT * FROM {0}_summary".format(sssp_table))
+		vertex_id = summary[0]['vertex_id']
+		if vertex_id == "NULL":
+			vertex_id = "id"
+
+		grouping_cols = summary[0]['grouping_cols']
+		if grouping_cols != "NULL":
+			glist = split_quoted_delimited_str(grouping_cols)
+			select_grps = _grp_from_table(sssp_table,glist) + " , "
+
+		plpy.execute("""
+			CREATE TABLE {path_table} AS
+				SELECT {select_grps} ''::text as path
+				FROM {sssp_table}
+				LIMIT 0
+			""".format(**locals()))
+
+		plan = """ PREPARE {plan_name} (int) AS
+			SELECT parent FROM {sssp_table}
+			WHERE {vertex_id} = $1 AND parent IS NOT NULL {gsql} LIMIT 1
+			"""
+
+		if grouping_cols == "NULL":
+			plpy.execute(plan.format(**locals()))
+			path_flag = _path_helper(plan_name,dest_vertex,path_table,"")
 
-	if parent.nrows() == 0:
-		plpy.error(
-			"Graph SSSP: Vertex {0} is not present in the sssp table {1}".
-			format(dest_vertex,sssp_table))
-
-	while 1:
-		parent = parent[0]['parent']
-		if parent == cur:
-			ret.reverse()
-			return ret
 		else:
-			ret.append(parent)
-			cur = parent
-		parent = plpy.execute(sql.format(**locals()))
+			gvalues = plpy.execute(
+				""" SELECT {grouping_cols} FROM {sssp_table}
+					GROUP BY {grouping_cols}""".format(**locals()))
+
+			for i in gvalues:
+				cur = dest_vertex
+				gcheck = []
+				gset = []
+				for j in glist:
+					gcheck.append("{0} = {1}".format(j,i[j]))
+					gset.append("{0}".format(i[j]))
+				gsql = " AND " + " AND ".join(gcheck)
```

Comment:
Try using list comprehension here as well, it is supposed
to be faster than append.